### PR TITLE
Add locks to param config

### PR DESCRIPTION
**Pull request by terencechain**
**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

We should add lock around param config usages. Some tests that I wrote in #8094 failed race because of it. 

Race log:
```

WARNING: DATA RACE
--
  | Write at 0x000002eb6aa0 by goroutine 44:
  | github.com/prysmaticlabs/prysm/shared/params.OverrideBeaconConfig()
  | shared/params/config.go:133 +0x145
  | github.com/prysmaticlabs/prysm/beacon-chain/blockchain.TestService_ReceiveBlock()
  | beacon-chain/blockchain/receive_block_test.go:35 +0x135
  | testing.tRunner()
  | GOROOT/src/testing/testing.go:1123 +0x202
  |  
  | Previous read at 0x000002eb6aa0 by goroutine 32:
  | github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).HeadRoot()
  | beacon-chain/blockchain/chain_info.go:125 +0x3f4
  | github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).saveHead()
  | beacon-chain/blockchain/head.go:86 +0x117
  | github.com/prysmaticlabs/prysm/beacon-chain/blockchain.TestHeadBlock_DataRace.func1()
  | beacon-chain/blockchain/chain_info_norace_test.go:46 +0x88
```

**Which issues(s) does this PR fix?**

Discovered in #8094

**Other notes for review**

### CODE REVIEW

File: shared/params/config.go [hunk 1]
```diff
@@ -130,11 +134,15 @@ func BeaconConfig() *BeaconChainConfig {
...
 // return this new configuration.
 func OverrideBeaconConfig(c *BeaconChainConfig) {
+	lock.Lock()
+	defer lock.Unlock()
 	beaconConfig = c
 }
 
 // Copy returns a copy of the config object.
 func (c *BeaconChainConfig) Copy() *BeaconChainConfig {
+	lock.Lock()
```
Comment (prestonvanloon) [inline shared/params/config.go:144]: This should be RLock?

Comment (terencechain) [reply shared/params/config.go:144]: I agree. This is tricker than I thought. Here is the problem: 
```go
func UseToledoConfig() {
	lock.Lock()                                     // We lock here
	defer lock.Unlock()
	beaconConfig = ToledoConfig()
}

func ToledoConfig() *BeaconChainConfig {
	cfg := MainnetConfig().Copy()
	cfg.MinGenesisTime = 1605009600
	return cfg
}

func (c *BeaconChainConfig) Copy() *BeaconChainConfig {
	lock.Lock()                                     // We'll dead lock here 
	defer lock.Unlock()
	config, ok := deepcopy.Copy(*c).(BeaconChainConfig)
	if !ok {
		config = *beaconConfig
	}
	return &config
}
```

Comment (terencechain) [reply shared/params/config.go]: Ok, fixed. I refactored to the following:

```go
func UseToledoConfig() {
	lock.Lock()                                     // We lock here
	defer lock.Unlock()
	beaconConfig = toledoConfig()
}

// This defines the config for the
// Toledo testnet.
// This is the lock free version.
func toledoConfig() *BeaconChainConfig {
	cfg := MainnetConfig().copy()
	cfg.MinGenesisTime = 1605009600
	cfg.GenesisDelay = 86400
	cfg.GenesisForkVersion = []byte{0x00, 0x70, 0x1E, 0xD0}
	cfg.NetworkName = "toledo"
	cfg.SecondsPerETH1Block = 14
	return cfg
}

// This returns a copy of the config object.
// This is the lock free version.
func (c *BeaconChainConfig) copy() *BeaconChainConfig {
	config, ok := deepcopy.Copy(*c).(BeaconChainConfig)
	if !ok {
		config = *beaconConfig
	}
	return &config
}
```

Please take a look @prestonvanloon

**Review by nisdas — COMMENTED** [review]
I dont know if this is such a good idea. Adding in a lock for our config, basically puts it in the hot path for everything. Nearly every method in our core package/services will be accessing the same config at the same time. They will all also have to acquire/release the lock here, bringing additional overhead.

**Comment by terencechain** [comment]
> I dont know if this is such a good idea. Adding in a lock for our config, basically puts it in the hot path for everything. Nearly every method in our core package/services will be accessing the same config at the same time. They will all also have to acquire/release the lock here, bringing additional overhead.  

For the read locks, concurrent reads can happen with negligible performance impact.

This lock is proposed because there's a race failure with when we overwrite config in tests, in the future, someone could also try this at run time, I do prefer safety.

**Comment by nisdas** [comment]
> For the read locks, concurrent reads can happen with negligible performance impact.

Are there any benchmarks to prove this ?

> This lock is proposed because there's a race failure with when we overwrite config in tests, in the future, someone could also try this at run time, I do prefer safety.

Wouldn't that be a test setup issue then ? From the failure it looks like its because you have multiple tests running at the same time, shouldn't tests not be run in parallel then.

**Comment by terencechain** [comment]
> Are there any benchmarks to prove this ?

Can certainly write a benchmark to further prove this.

> Wouldn't that be a test setup issue then ? From the failure it looks like its because you have multiple tests running at the same time, shouldn't tests not be run in parallel then.

It's a test setup issue for now, but the worry here is it lingers long enough to become a run time / production issue. Imagine a simple code change that I made and was able to trigger this race failure with one test. This params package does not look safe to me without proper locking

**Comment by terencechain** [comment]
Here's the benchmark.
```go
func Benchmark_GetConfig(b *testing.B) {
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			if params.BeaconConfig().SlotsPerEpoch != 32 {
				b.Fatal("Incorrect SlotsPerEpoch")
			}
		}
	})
}
```

Without read lock:
pkg: github.com/prysmaticlabs/prysm/shared/params
Benchmark_GetConfig
Benchmark_GetConfig-12    	1000000000	         0.165 ns/op
PASS

With read lock:
pkg: github.com/prysmaticlabs/prysm/shared/params
Benchmark_GetConfig
Benchmark_GetConfig-12    	30282727	        34.5 ns/op
PASS

About 30ns slower. Question now is whether 30ns is an acceptable trade off for safety

**Comment by nisdas** [comment]
> It's a test setup issue for now, but the worry here is it lingers long enough to become a run time / production issue. Imagine a simple code change that I made and was able to trigger this race failure with one test, how likely it could happen again in the future. This params package does not look safe to me without proper locking

So the fundamental issue we must address is that do we expect to write to/initialize the config while the node is running normally (after node initialization) ? If there is the possibility that it can happen while the node is up and running normally and we do try to modify the config, we will need a lock here. But in the event that its possible for something like that to happen, we probably need to re-check our assumptions. We assume that the config is a static object for the most part, if it can be continuously modified in runtime it means a lot of things that we assume when running our services don't hold. 

> Without read lock:
>
>pkg: github.com/prysmaticlabs/prysm/shared/params
Benchmark_GetConfig
Benchmark_GetConfig-12    	1000000000	         0.165 ns/op
PASS
With read lock:

> pkg: github.com/prysmaticlabs/prysm/shared/params
Benchmark_GetConfig
Benchmark_GetConfig-12    	30282727	        34.5 ns/op
PASS
About 30ns slower. Question now is whether 30ns is an acceptable trade off for safety

Thats a pretty big difference, even though this is only 1 value it all eventually adds up . Probably a better benchmark is to benchmark state transitions with/without the config. Since that is our most important code path to be concerned of.

**Comment by terencechain** [comment]
> So the fundamental issue we must address is that do we expect to write to/initialize the config while the node is running normally (after node initialization)

It's not so much on expection, it's more on allowence. We can say that we don't expect one to modify config during run time but anyone can make a mistake doing it and we can miss reviewing it. The best solution us don't allow config overwritr and treat config as static file. The next best solution is use a lock. Do nothing is not good long term imo

**Comment by terencechain** [comment]
As promised, here's some benchmark numbers for full state transtion.

Without locking:
BenchmarkExecuteStateTransition_FullProcess-12          18  66811454 ns/op

With locking:
BenchmarkExecuteStateTransition_FullProcess-12    	      16	  67013178 ns/op

It's about 0.3% worse than before

Source:
```go
func BenchmarkExecuteStateTransition_FullProcess(b *testing.B) {
	beaconState, privKeys := testutil.DeterministicGenesisState(b, 100)

	eth1Data := &ethpb.Eth1Data{
		DepositCount: 100,
		DepositRoot:  bytesutil.PadTo([]byte{2}, 32),
		BlockHash:    make([]byte, 32),
	}
	require.NoError(b, beaconState.SetSlot(params.BeaconConfig().SlotsPerEpoch-1))
	e := beaconState.Eth1Data()
	e.DepositCount = 100
	require.NoError(b, beaconState.SetEth1Data(e))
	bh := beaconState.LatestBlockHeader()
	bh.Slot = beaconState.Slot()
	require.NoError(b, beaconState.SetLatestBlockHeader(bh))
	require.NoError(b, beaconState.SetEth1DataVotes([]*ethpb.Eth1Data{eth1Data}))

	require.NoError(b, beaconState.SetSlot(beaconState.Slot()+1))
	epoch := helpers.CurrentEpoch(beaconState)
	randaoReveal, err := testutil.RandaoReveal(beaconState, epoch, privKeys)
	require.NoError(b, err)
	require.NoError(b, beaconState.SetSlot(beaconState.Slot()-1))

	nextSlotState, err := state.ProcessSlots(context.Background(), beaconState.Copy(), beaconState.Slot()+1)
	require.NoError(b, err)
	parentRoot, err := nextSlotState.LatestBlockHeader().HashTreeRoot()
	require.NoError(b, err)
	proposerIdx, err := helpers.BeaconProposerIndex(nextSlotState)
	require.NoError(b, err)
	block := testutil.NewBeaconBlock()
	block.Block.ProposerIndex = proposerIdx
	block.Block.Slot = beaconState.Slot() + 1
	block.Block.ParentRoot = parentRoot[:]
	block.Block.Body.RandaoReveal = randaoReveal
	block.Block.Body.Eth1Data = eth1Data

	stateRoot, err := state.CalculateStateRoot(context.Background(), beaconState, block)
	require.NoError(b, err)

	block.Block.StateRoot = stateRoot[:]

	sig, err := testutil.BlockSignature(beaconState, block.Block, privKeys)
	require.NoError(b, err)
	block.Signature = sig.Marshal()

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		copied := beaconState.Copy()
		_, err = state.ExecuteStateTransition(context.Background(), copied, block)
		require.NoError(b, err)
	}
}
```

**Comment by terencechain** [comment]
closing this and exploring better solutions

**Comment by dshulyak** [comment]
if 30ns of overhead is your primary concern and updates are rare you can use this pattern with atomics:

```go
import (
	"sync/atomic"
	"unsafe"
)

var cfg unsafe.Pointer = unsafe.Pointer(&Config{})

type Config struct {
	value string
}

func getConfig() *Config {
	return (*Config)(atomic.LoadPointer(&cfg))
}

func updateConfig(f func(c *Config)) {
	for {
		c1 := getConfig()
		c2 := &Config{value: c1.value} // copy
		f(c2)
		if atomic.CompareAndSwapPointer(&cfg, unsafe.Pointer(c1), unsafe.Pointer(c2)) {
			return
		}
	}
}
```

getConfig has no overhead at all on x86 (will probably have insignificant on systems with weaker memory model)

**Pull request closed without merging**